### ISSUE TITLE
Line endings are forced to be UNIX endings \n

### ISSUE THREAD

**Issue by thecheeselover** [square/javapoet#600]
There should be a way to switch between LF and CR/LF endings but the \n endings are hardcoded within the JavaFile class.

**Comment by topalavlad** [comment]
The hardcoded new line endings `\n` seem to be used in ~9 files.
I think even though it's not the prettiest thing, it would be easiest for the parameter to be static for example in the `Util` class or somewhere similar. Any opinions on this?

**Comment by thecheeselover** [comment]
That or if the writer could have a default object that returns a line ending, so that we could also supply another object that returns a desired line ending.

**Comment by the-end-sea-of-the-world** [comment]
Hello, I would like to work on this.

**Comment by JakeWharton** [comment]
Can you describe your motivation for wanting different line endings? Any modern editor will be able to handle newline-delimited lines automatically and `javac` doesn't care. 

Changing the line endings based on platforms makes testing dramatically harder as you can no longer compare golden files/strings directly to the output. Even worse, build systems like Bazel and Gradle which hash inputs will get cache misses when developers are working on different operating systems preventing distributed caching from working and forcing needless recompilations.

If we were to add this (and that's a huge _if_, as I currently see no justification for it) then it would probably have to be a configurable value on each `FileSpec`. Definitely not some static value and definitely cannot be inferred from the platform.

**Comment by jbeckers** [comment]
This is probably a weak argument, but... :-)
We use javapoet to generate code that ends up in an IntelliJ project that has Windows line endings.
When viewing the generated code in IJ, it complains that the file's line endings are wrong.
Not a big deal, but annoying none the less.

**Comment by Inego** [comment]
JavaFile has `writeToPath` that accepts a charset.
If charsets are supported then line endings also should be.

**Comment by lyahim** [comment]
Hi all,

Please think about to implement this feature because on Windows environment line endings CRLF.
Java is a platform independent language and the application should be same. If a java source generator don't support System.lineSeparator() the applications which use that can't be platform independent.

**Comment by JakeWharton** [comment]
LF-delimited files compile fine on Windows, and have for the entirety of this library's existence. Moreover, an exact, byte-for-byte output is essential for distributed build caches to work correctly with tools like Gradle. At this point, we aren't changing the library behavior.

**Comment by lyahim** [comment]
You are right, generated files are correctly compiled on Windows too but Git on Windows by default try to force CRLF which cause differences. Of course it could handle with some effort.
I agree with that not a good idea to change the default behavior of the generator but would be good extend with a feature to give a possibility to change it. For example public void writeTo(Path directory, Charset charset, String lineSeparator).
With this the old generated files would be same as before and future clients can decide to use an other separator.
I saw in the code that this update should many code changes when I have time I create a PR to show what I mean.

**Comment by klopfdreh** [comment]
Please don't change the code but use `.gitattributes` - the code itself uses `\n` all the time which in java environments should be fine.

The content of the `.gitattributes`:
```plain
# See: https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings
* text eol=lf
```

**Comment by Inego** [comment]
> Please don't change the code but use `.gitattributes` - the code itself uses `\n` all the time which in java environments should be fine.
> 
> The content of the `.gitattributes`:
> 
> ```
> # See: https://docs.github.com/en/get-started/getting-started-with-git/configuring-git-to-handle-line-endings
> * text eol=lf
> ```
> 
> > text eol=lf Git will always convert line endings to LF on checkout. You should use this for files that must keep LF endings, even on Windows.

This issue/request doesn't mention git (a user might have no VCS at all) — it's about controlling the textual output of the code generation.
